File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/manual.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.Servo;

import java.util.Arrays;

@TeleOp(name="Manual", group="Linear OpMode")
public class manual extends LinearOpMode {

    Motor[] motors;

    @Override
    public void runOpMode() {
        this.init_hardware();
        waitForStart();
        while (opModeIsActive()) {
            this.move_drivetrain();
        }
    }

    void move_drivetrain() {
        double axial = -gamepad1.left_stick_y;
        double lateral = gamepad1.right_stick_x;
        double yaw = gamepad1.left_stick_x;
        telemetry.update();

        motors[0].power = axial + yaw + lateral;
        motors[1].power = axial - yaw + lateral;
        motors[2].power = axial - yaw - lateral;
        motors[3].power = axial + yaw - lateral;

        double max = Arrays.stream(motors).mapToDouble(motor -> Math.abs(motor.power)).max().orElse(1);
        for (Motor motor : motors) {
            double scaledPower = motor.power;
            if (max > 1.0) {
                scaledPower /= max;
            }
            motor.drive.setPower(scaledPower*0.2);
        }
    }

    void init_hardware() {
        motors = new Motor[]{
                new Motor("left_front_drive", DcMotor.Direction.FORWARD),
                new Motor("left_back_drive", DcMotor.Direction.FORWARD),
                new Motor("right_front_drive", DcMotor.Direction.REVERSE),
                new Motor("right_back_drive", DcMotor.Direction.REVERSE),
        };
    }

    class Motor {
        DcMotor drive;
        double power;

        Motor(String name, DcMotorSimple.Direction direction) {
            this.drive = hardwareMap.get(DcMotor.class, name);
            this.drive.setDirection(direction);
            this.drive.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
            this.drive.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Manual.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.util.ElapsedTime
import org.firstinspires.ftc.teamcode.robotParts.Constants.MANUAL_MULTIPLIER
import org.firstinspires.ftc.teamcode.robotParts.OdometryDrivetrain
import org.firstinspires.ftc.teamcode.robotParts.Shooter
import org.firstinspires.ftc.teamcode.robotParts.TransferIntake
import kotlin.math.abs

enum class RobotState {INTAKE, SHOOTER_SPIN_UP, SHOOTER_ON}

/**
 * CONTROLS:
 * LEFT / RIGHT joystick -> movement
 * LEFT BUMPER to go into SHOOTER mode (SPIN UP) - only do this if you want to shoot from there
 * RIGHT BUMPER to enter INTAKE mode
 * Triggers -> intake (forwards / back)
 * DPAD UP/DOWN to move the goal 10cm FURTHER/NEARER (virtually)
 **/

abstract class Manual(private val orientation: Double, private val tag: Int) : LinearOpMode() {
    private lateinit var drivetrain: OdometryDrivetrain
    private lateinit var shooter: Shooter
    private lateinit var transferIntake: TransferIntake
    private var robotState = RobotState.INTAKE
    private var timeToEnd = 0.0
    private val timer = ElapsedTime()
    private val atSpeed = ElapsedTime()
    private val dpadTimer = ElapsedTime()
    override fun runOpMode() {
        waitForStart()
        drivetrain = OdometryDrivetrain(hardwareMap, 0.0,0.0,orientation)
        shooter = Shooter(hardwareMap, tag)
        transferIntake = TransferIntake(hardwareMap)
        while (opModeIsActive()) {
            /** expected field centric control **/
            drivetrain.driveManual(
                gamepad1.left_stick_x*abs(gamepad1.left_stick_x).toDouble(),
                -gamepad1.left_stick_y*abs(gamepad1.left_stick_y).toDouble(),
                gamepad1.right_stick_x*abs(gamepad1.right_stick_x).toDouble(), MANUAL_MULTIPLIER
            )
            if (gamepad1.left_bumper && timer.milliseconds() > timeToEnd) {
                robotState = RobotState.SHOOTER_SPIN_UP
                transferIntake.prepShooter()
                shooter.turnOnShooter()
                timeToEnd = timer.milliseconds() + 500
            }
            if (gamepad1.right_bumper && timer.milliseconds() > timeToEnd) {
                robotState = RobotState.INTAKE
                transferIntake.shoot(false)
                shooter.turnOffShooter()
                timeToEnd = timer.milliseconds() + 200
            }
            if (gamepad1.dpad_up && dpadTimer.milliseconds() > 200) {
                shooter.turretOffset += 0.1
                dpadTimer.reset()
            }
            if (gamepad1.dpad_down && dpadTimer.milliseconds() > 200) {
                shooter.turretOffset -= 0.1
                dpadTimer.reset()
            }
            if (robotState == RobotState.INTAKE) { transferIntake.intake(gamepad1.left_trigger - gamepad1.right_trigger) }
            if (!shooter.atSpeed) atSpeed.reset()
            if (atSpeed.milliseconds() > 500 && (robotState == RobotState.SHOOTER_SPIN_UP) && shooter.canShoot() && timer.milliseconds() > timeToEnd) {
                robotState = RobotState.SHOOTER_ON
                transferIntake.shoot(true)
            }
            transferIntake.update()
            shooter.moveTurret()
            shooter.spin()
            telemetry.addLine(shooter.getData())
            telemetry.addLine(transferIntake.getData())
            telemetry.addLine(drivetrain.getData())
            telemetry.update()
        }
    }
}